File: frontend/types/workflow.ts
export type WorkflowTriggerType =
  | "ticket_created"
  | "ticket_status_changed"
  | "ticket_priority_changed"
  | "ticket_assigned"
  | "ticket_unassigned"
  | "ticket_comment_added"
  | "ticket_sentiment_changed"
  | "ticket_idle";

export type WorkflowConditionField =
  | "status"
  | "priority"
  | "sentiment"
  | "team_id"
  | "assigned_to"
  | "is_unassigned"
  | "subject"
  | "description"
  | "hours_since_updated"
  | "hours_since_created";

export type WorkflowOperator =
  | "equals"
  | "not_equals"
  | "in"
  | "contains"
  | "greater_than"
  | "less_than"
  | "is_empty"
  | "is_not_empty";

export type WorkflowActionType =
  | "assign_team"
  | "assign_user"
  | "unassign"
  | "set_priority"
  | "set_status"
  | "add_internal_note";

export type ConditionLogic = "all" | "any";

export interface WorkflowCondition {
  id?: string;
  field: WorkflowConditionField;
  operator: WorkflowOperator;
  value: string | null;
  order: number;
}

export interface WorkflowAction {
  id?: string;
  action_type: WorkflowActionType;
  params: Record<string, string>;
  order: number;
}

export interface WorkflowRule {
  id: string;
  name: string;
  description: string | null;
  trigger_type: WorkflowTriggerType;
  condition_logic: ConditionLogic;
  is_active: boolean;
  run_order: number;
  created_by: string;
  created_at: string;
  updated_at: string;
  conditions: WorkflowCondition[];
  actions: WorkflowAction[];
}

export interface WorkflowRuleInput {
  name: string;
  description?: string | null;
  trigger_type: WorkflowTriggerType;
  condition_logic: ConditionLogic;
  is_active: boolean;
  run_order: number;
  conditions: WorkflowCondition[];
  actions: WorkflowAction[];
}

export interface WorkflowExecutionLog {
  id: string;
  rule_id: string;
  ticket_id: string;
  trigger_type: WorkflowTriggerType;
  success: boolean;
  error_message: string | null;
  actions_summary: { actions?: string[] };
  created_at: string;
}

File: backend/app/schemas/workflow.py
import uuid
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict

from app.workflows.enums import (
    WorkflowTriggerType,
    WorkflowConditionField,
    WorkflowOperator,
    WorkflowActionType,
    ConditionLogic,
)


class WorkflowConditionCreate(BaseModel):
    field: WorkflowConditionField
    operator: WorkflowOperator
    value: str | None = None
    order: int = 0


class WorkflowConditionOut(WorkflowConditionCreate):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID


class WorkflowActionCreate(BaseModel):
    action_type: WorkflowActionType
    params: dict = Field(default_factory=dict)
    order: int = 0


class WorkflowActionOut(WorkflowActionCreate):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID


class WorkflowRuleCreate(BaseModel):
    name: str
    description: str | None = None
    trigger_type: WorkflowTriggerType
    condition_logic: ConditionLogic = ConditionLogic.ALL
    is_active: bool = True
    run_order: int = 0
    conditions: list[WorkflowConditionCreate] = Field(default_factory=list)
    actions: list[WorkflowActionCreate] = Field(min_length=1)


class WorkflowRuleUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    trigger_type: WorkflowTriggerType | None = None
    condition_logic: ConditionLogic | None = None
    is_active: bool | None = None
    run_order: int | None = None
    conditions: list[WorkflowConditionCreate] | None = None
    actions: list[WorkflowActionCreate] | None = None


class WorkflowRuleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    description: str | None
    trigger_type: WorkflowTriggerType
    condition_logic: ConditionLogic
    is_active: bool
    run_order: int
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime
    conditions: list[WorkflowConditionOut]
    actions: list[WorkflowActionOut]


class WorkflowExecutionLogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    rule_id: uuid.UUID
    ticket_id: uuid.UUID
    trigger_type: WorkflowTriggerType
    success: bool
    error_message: str | None
    actions_summary: dict
    created_at: datetime